# MaveDB

MaveDB is a public repository for datasets from Multiplexed Assays of Variant Effect (MAVEs), such as those generated by deep mutational scanning (DMS) or massively parallel reporter assay (MPRA) experiments. Despite the importance of MAVE data for basic and clinical research, there is no standard resource for their discovery and distribution. Here, we present MaveDB (https://www.mavedb.org), a public repository for large-scale measurements of sequence variant impact, designed for interoperability with applications to interpret these datasets.

MaveDB is open-source, released under the AGPLv3 license.

MaveDB is hosted by the Fowler Lab in the Department of Genome Sciences at the University of Washington. It is supported and developed by the University of Washington, the Walter and Eliza Hall Institute of Medical Research, and the Brotman Baty Institute.

## Data Organization

MaveDB has three dataset types: Score Set, Experiment, and Experiment Set. All score, count, and target information is stored in score sets. Experiment sets and experiments are used for organization and metadata.

The dataset types are organized hierarchically. Each experiment set can contain multiple experiments and each experiment can contain multiple score sets.

## Score Sets

To capture the structure of real-world study designs, MaveDB is organized hierarchically into score sets, experiments, and experiment sets. Score sets, the most basic unit of organization, contain the variant effect scores and additional metadata such as target sequence information and detailed methods. Each variant effect score is a numeric value. 

## Accession Number Formats

MaveDB accession numbers use the URN (Uniform Resource Name) format. These accession numbers have a hierarchical structure that reflects the relationship between experiment sets, experiments, score sets, and individual variants in MaveDB.

## Studies

MaveDB's variant effect scores are captured from two separate studies, involving specific genes and accession numbers. The following provides information to which genes are included in each study, and methodology behind calculating the scores. 

### Calmodulin yeast complementation
* Includes genes: CALM1, SUMO1, UBE2I, and TPK1
* Method: A Deep Mutational Scan of Calmodulin (CALM1) using functional complementation in yeast was performed using DMS-TileSeq and a machine-learning method was used to impute the effects of missing variants and refine measurements of lower confidence.

### VAMP-seq
* Includes genes: PTEN and TPMT
* Method: Barcoded variant libraries were created using inverse PCR. Barcodes were associated with full-length variants using Pacific Biosciences SMRT sequencing to generate long reads. Each protein variant is fused to EGFP so that its abundance can be tracked using fluorescence. Cells were sorted into bins using FACS based on the ratio of EGFP to mCherry.

## References

Matreyek, Kenneth A., et al. "Multiplex assessment of protein variant abundance by massively parallel sequencing." Nature genetics 50.6 (2018): 874-882.

Weile, Jochen, et al. "A framework for exhaustively mapping functional missense variants." Molecular systems biology 13.12 (2017): 957.

Information from https://www.mavedb.org/
