# MITOMAP: A human mitochondrial genome database

The MITOMAP database of human mitochondrial DNA (mtDNA) information has been an important resource for information about the human mitochondrial DNA (mtDNA) for researchers, clinicians and genetic counselors for the past twenty-five years. Essential information about the mitochondrial reference sequence is provided along with an extensive compilation of mtDNA variants. The MITOMAP curators search research literature for published reports of mitochondrial DNA variants and index those variants in the database. Those variants which are reported as having possible association with disease are noted. A new addition to MITOMAP is the inclusion of data from full-length human mtDNA sequences in GenBank.

## MitoTIP Scores

The pathogenicity scoring from MitoTIP is intended as a starting point for analysis of newly observed tRNA variants. It is based only on database frequencies, the nature of the nucleotide change (transversion/transition/deletion), and conservation scoring. No clinical histories, heteroplasmy data, or functional studies are included. Variants must be further evaluated by the end user in the context of the actual patient heteroplasmy, and the heteroplasmy seen in affected and unaffected family members. Other additional lines of evidence are suggested below.
MitoTIP integrates multiple sources of information to provide a prediction for the likelihood that novel single nucleotide variants or deletions in tRNA-encoding sequences would cause disease. The sources of information used include:
* GenBank deposited mitochondrial sequence
* Annotations of pathogenicity from MITOMAP
* Conservation across species
* The position of the variant within the tRNA
* Whether the nucleotide change is a transition or transversion

Each possible nucleotide change was scored and the scores have been interpreted within quartiles. Current MitoTIP scores range from -5.9 to 21.8, with a median score of 12.7. MitoTIP scores will be recalculated on a regular basis.
MitoTIP's scoring of tRNA variants is embedded into Mitomaster and may be viewed by using Mitomaster's SNV query or Sequence query tools.

Information from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4257604/ and https://mitomap.org/MITOMAP/MitoTipInfo



