# 23andMe Converter

Converts typical 23andMe text output, which can be downloaded by following the instructions [here](https://customercare.23andme.com/hc/en-us/articles/212196868-Accessing-Your-Raw-Genetic-Data).



Files typically look like this:
```
# This data file generated by 23andMe at: XX XXX XXX XX:XX:XX XXXX
#
# This file contains raw genotype data, including data that is not used in 23andMe reports.
# This data has undergone a general quality review however only a subset of markers have been 
# individually validated for accuracy. As such, this data is suitable only for research, 
# educational, and informational use and not for medical or other use.
# 
# Below is a text version of your data.  Fields are TAB-separated
# Each line corresponds to a single SNP.  For each SNP, we provide its identifier 
# (an rsid or an internal id), its location on the reference human genome, and the 
# genotype call oriented with respect to the plus strand on the human reference sequence.
# We are using reference human assembly build 37 (also known as Annotation Release 104).
# Note that it is possible that data downloaded at different times may be different due to ongoing 
# improvements in our ability to call genotypes. More information about these changes can be found at:
# https://you.23andme.com/p/XXXXXX/tools/data/download/
# 
# More information on reference human assembly builds:
# https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.13/
#
# rsid	chromosome	position	genotype
rs548049170	1	69869	TT
rs13328684	1	74792	--
```

The header lines must be included for the file to be recognized as a 23andMe input.

As of 2020/1/28, most 23andMe files will be on the hg19/GRCh37 genome. Be sure to specify the correct genome when submitting a job. Use the `-l/--liftover hg19` option in the cmd line, or the "Genome" dropdown at the top of the submission section of the GUI.
